src/main/scala/net/ripe/rpki/publicationserver/store/postgresql/PgStore.scala

```diff
@@ -0,0 +1,248 @@
+package net.ripe.rpki.publicationserver.store.postresql
+
+import java.net.URI
+
+import net.ripe.rpki.publicationserver.Binaries.Bytes
+import net.ripe.rpki.publicationserver._
+import net.ripe.rpki.publicationserver.metrics.Metrics
+import net.ripe.rpki.publicationserver.model._
+import org.flywaydb.core.Flyway
+import org.json4s._
+import org.json4s.native.JsonMethods._
+import scalikejdbc._
+
+
+case class RollbackException(val error: BaseError) extends Exception
+
+class PgStore(val pgConfig: PgConfig) extends Hashing with Logging {
+
+  def inRepeatableReadTx[T](f: DBSession => T): T = {
+    DB(ConnectionPool.borrow())
+      .isolationLevel(IsolationLevel.RepeatableRead)
+      .localTx(f)
+  }
+
+  def clear(): Unit = DB.localTx { implicit session =>
+    sql"TRUNCATE TABLE object_log CASCADE".update().apply()
+    sql"TRUNCATE TABLE objects CASCADE".update().apply()
+    sql"TRUNCATE TABLE versions CASCADE".update().apply()
+  }
+
+  def getState = DB.localTx { implicit session =>
+    sql"SELECT * FROM current_state"
+      .map { rs =>
+        val hash = Hash(rs.string(1))
+        val uri = URI.create(rs.string(2))
+        val clientId = ClientId(rs.string(3))
+        val bytes = Bytes.fromStream(rs.binaryStream(4))
+        uri -> (bytes, hash, clientId)
+      }
+      .list()
+      .apply()
+      .toMap
+  }
+
+  def getLog = DB.localTx { implicit session =>
+    sql"""SELECT operation, url, old_hash, content
+         FROM current_log"""
+      .map { rs =>
+        val operation = rs.string(1)
+        val uri = URI.create(rs.string(2))
+        val hash = rs.stringOpt(3).map(Hash)
+        val bytes = rs.binaryStreamOpt(4).map(Bytes.fromStream)
+        (operation, uri, hash, bytes)
+      }
+      .list()
+      .apply()
+  }
+
+  def readState(f: (URI, Hash, Bytes) => Unit)(implicit session: DBSession) = {
+    session.fetchSize(200)
+    sql"SELECT url, hash, content FROM current_state"
+      .foreach { rs =>
+        val uri = URI.create(rs.string(1))
+        val hash = Hash(rs.string(2))
+        val bytes = Bytes.fromStream(rs.binaryStream(3))
+        f(uri, hash, bytes)
```

Comment:
`readDelta` is ordered by url, would it hurt if we sort these objects in current state when writing Snapshots?  Helps comparing snapshots if you happen to have multiple publication servers.

Reply:
Good point, will update.